sway-core/src/asm_generation/from_ir.rs

```diff
@@ -2090,98 +2102,195 @@ impl<'ir> AsmBuilder<'ir> {
+                    comment: "initialise aggregate field".into(),
+                    owning_span: span,
+                });
+
+                value_size_in_words
+            }
+            Type::Array(aggregate) => {
+                match (&constant.value, &self.context.aggregates[aggregate.0]) {
+                    (ConstantValue::Array(items), AggregateContent::ArrayType(element_type, _)) => {
+                        // Recurse for each item, accumulating the field offset and the final size.
+                        items.iter().fold(0, |local_offs, item| {
+                            local_offs
+                                + self.initialise_constant_memory(
+                                    item,
+                                    element_type,
+                                    start_reg,
+                                    offs_in_words + local_offs,
+                                    span.clone(),
+                                )
+                        })
+                    }
+                    _ => unreachable!("Inconsistent types for constant initialisation"),
+                }
+            }
+            Type::Struct(aggregate) => {
+                match (&constant.value, &self.context.aggregates[aggregate.0]) {
+                    (ConstantValue::Struct(items), AggregateContent::FieldTypes(field_tys)) => {
+                        // Recurse for each item, accumulating the field offset and the final size.
+                        items.iter().zip(field_tys.iter()).fold(
+                            0,
+                            |local_offs, (item, field_tys)| {
+                                local_offs
+                                    + self.initialise_constant_memory(
+                                        item,
+                                        field_tys,
+                                        start_reg,
+                                        offs_in_words + local_offs,
+                                        span.clone(),
+                                    )
+                            },
+                        )
+                    }
+                    _ => unreachable!("Inconsistent types for constant initialisation"),
+                }
+            }
+            Type::Union(_) => {
+                // If the literal we're trying to initialise with is smaller than than the actual
+                // size of the union, then a padding of zeros is required. Calculate the size of
+                // the padding and set the appropriate bytes to zero.
+                let constant_size = ir_type_size_in_bytes(self.context, &constant.ty);
+                assert!(constant_size <= value_size);
+                let padding_size = value_size - constant_size;
+                let padding_size_in_words = size_bytes_in_words!(padding_size);
+
+                if padding_size > 0 {
+                    // Store padding of zeros and then store the value itself
```

Comment:
So, do you want to put the padding first?  While the values _are_ big endian, when the variant is smaller we'll expect it to butt up against the tag won't we?  Or have you changed the way that works elsewhere?

Reply:
I feel like left padding is more intuitive? But also makes the math a bit more annoying to follow so I don't really have a strong opinion here. This should definitely be in the spec somewhere because it affects external contexts.

Reply:
Sadly, the example in the spec does not tell us much about enums with enums that have different sizes: https://github.com/FuelLabs/fuel-specs/blob/master/specs/protocol/abi.md#sum-types-enums

Reply:
Created an issue here: https://github.com/FuelLabs/fuel-specs/issues/339